small fixes + add automerge for TemplateClustering

jenniferColonell opened the pull request:
A repeat of my previous pull request, now in files with correct line endings. Please just close the previous pull request.

aliddell commented on +jrclust/+curate/@CurateController/updateFigWav.m:
```diff
@@ -3,8 +3,10 @@ function updateFigWav(obj)
         return;
     end
-
+    fprintf('Number of clusters at update FigWav: %d\n', obj.hClust.nClusters);
```
What do you use this for? Could it be logged instead?

aliddell commented on +jrclust/+curate/@CurateController/updateFigWav.m:
```diff
@@ -3,8 +3,10 @@ function updateFigWav(obj)
+    fprintf('Number of clusters at update FigWav: %d\n', obj.hClust.nClusters);
+    
     hFigWav = obj.hFigs('FigWav');
+    obj.showSubset = 1:obj.hClust.nClusters;
```
Why does this need to be explicitly reset?

aliddell commented on +jrclust/+sort/@DensityPeakClustering/DensityPeakClustering.m:
```diff
@@ -104,8 +104,13 @@
     end
 
     %% UTILITY METHODS
+%     methods (Access=protected, Hidden)
```
You can remove commented out lines here.

aliddell commented on +jrclust/+sort/@TemplateClustering/autoMerge.m:
```diff
@@ -1,4 +1,26 @@
-function autoMerge(obj)
+function success = autoMerge(obj)
```
This looks (almost) identical to the `autoMerge` in `DensityPeakClustering`. We can move them both to `Clustering` instead, so changes later don't have to be duplicated.

aliddell commented on +jrclust/+sort/@TemplateClustering/autoMerge.m:
```diff
@@ -1,4 +1,26 @@
+        break;
+    end
+end
+fprintf( 'size of waveform similarity matrix: \n')
```
Again, something I'd log rather than print. Or is this just debug output?

aliddell commented on +jrclust/+sort/@TemplateClustering/mergeBySim.m:
```diff
@@ -1,5 +1,42 @@
 function nMerged = mergeBySim(obj)
-    %MERGEBYSIM
-    nMerged = 0;
+
```
This one is also (almost) identical to `DensityPeakClustering.mergeBySim`. I suggest moving them both to `Clustering`.

aliddell commented on +jrclust/+sort/@TemplateClustering/mergeBySim.m:
```diff
@@ -1,5 +1,42 @@
-    %MERGEBYSIM
-    nMerged = 0;
+
+    fprintf('Template clustering mergeBySim\n');
```
Debug output?

aliddell commented on +jrclust/+utils/metaToStruct.m:
```diff
@@ -11,7 +11,7 @@
 
     fid = fopen(filename, 'r');
-    keysvals = textscan(fid, '%s%s', 'Delimiter', '=',  'ReturnOnError', 0);
+    keysvals = textscan(fid, '%[^=] = %[^\r\n]',  'ReturnOnError', 0);
```
Is this more robust? Did the other way fail on you in some way? Does it handle cases where the `=` isn't separated by spaces?

jenniferColonell commented:
Hi Alan,

My answers in red, below. Sorry I left in all those debugging comments!

Have removed and pushed up to my fork.

Thanks!

jennifer

________________________________
From: Alan Liddell <notifications@github.com>
Sent: Thursday, March 26, 2020 9:41 AM
To: JaneliaSciComp/JRCLUST <JRCLUST@noreply.github.com>
Cc: Colonell, Jennifer <colonellj@janelia.hhmi.org>; Author <author@noreply.github.com>
Subject: Re: [JaneliaSciComp/JRCLUST] small fixes + add automerge for TemplateClustering (#144)


@aliddell requested changes on this pull request.

I've got some changes and some questions for you. Feel free to counter-question!

________________________________

In +jrclust/+curate/@CurateController/autoMerge.m<https://urldefense.com/v3/__https://github.com/JaneliaSciComp/JRCLUST/pull/144*discussion_r398549258__;Iw!!Eh6p8Q!WpOR-hIelz5ipAykGehbdqfmrGnn9mgVrGlgcC5PmQMYIjbkYMsITIgfCVeHvCnZApcSEQ$>:

> @@ -4,9 +4,9 @@ function autoMerge(obj, maxUnitSim)
     if obj.isWorking
         jrclust.utils.qMsgBox('An operation is in progress.');
         return;
-    elseif ~isa(obj.hClust, 'jrclust.sort.DensityPeakClustering')
-        jrclust.utils.qMsgBox('Operation not supported for this type of clustering');
-        return;
+%     elseif ~isa(obj.hClust, 'jrclust.sort.DensityPeakClustering')


You can just delete these lines rather than commenting them out. If we need to we can recover them.


=> Removed commented out lines

________________________________

In +jrclust/+curate/@CurateController/updateFigWav.m<https://urldefense.com/v3/__https://github.com/JaneliaSciComp/JRCLUST/pull/144*discussion_r398564494__;Iw!!Eh6p8Q!WpOR-hIelz5ipAykGehbdqfmrGnn9mgVrGlgcC5PmQMYIjbkYMsITIgfCVeHvCkzE_dumw$>:

> @@ -3,8 +3,10 @@ function updateFigWav(obj)
     if ~obj.hasFig('FigWav')
         return;
     end
-
+    fprintf('Number of clusters at update FigWav: %d\n', obj.hClust.nClusters);


What do you use this for? Could it be logged instead?


=> Just diagnostic, can remove

________________________________

In +jrclust/+curate/@CurateController/updateFigWav.m<https://urldefense.com/v3/__https://github.com/JaneliaSciComp/JRCLUST/pull/144*discussion_r398566817__;Iw!!Eh6p8Q!WpOR-hIelz5ipAykGehbdqfmrGnn9mgVrGlgcC5PmQMYIjbkYMsITIgfCVeHvCnQUYrgtA$>:

>      hFigWav = obj.hFigs('FigWav');
+    obj.showSubset = 1:obj.hClust.nClusters;


Why does this need to be explicitly reset?


=> When running merge auto for imported ksort data, the number of clusters doesn't get updated properly (I think maybe this was true for standard merges as well -- but only in some interim version of Adrian's code). Let me know if the update properly belongs somewhere else. Without the update, I get this error during merge auto:


Index exceeds the number of array elements (54).

Error in plotMeanWaveforms (line 36)
        iSites = hCfg.siteNeighbors(:, hClust.clusterSites(iCluster));

Error in jrclust.views.plotFigWav (line 37)
        hFigWav = plotMeanWaveforms(hFigWav, hClust, maxAmp,channel_idx);

Error in jrclust.curate.CurateController/updateFigWav (line 10)
    jrclust.views.plotFigWav(hFigWav, obj.hClust, obj.maxAmp, obj.showSubset, obj.channel_idx);
________________________________

In +jrclust/+sort/@DensityPeakClustering/DensityPeakClustering.m<https://urldefense.com/v3/__https://github.com/JaneliaSciComp/JRCLUST/pull/144*discussion_r398567128__;Iw!!Eh6p8Q!WpOR-hIelz5ipAykGehbdqfmrGnn9mgVrGlgcC5PmQMYIjbkYMsITIgfCVeHvClS5Cdhxw$>:

> @@ -104,8 +104,13 @@
     end

     %% UTILITY METHODS
+%     methods (Access=protected, Hidden)


You can remove commented out lines here.

________________________________

In +jrclust/+sort/@TemplateClustering/autoMerge.m<https://urldefense.com/v3/__https://github.com/JaneliaSciComp/JRCLUST/pull/144*discussion_r398569098__;Iw!!Eh6p8Q!WpOR-hIelz5ipAykGehbdqfmrGnn9mgVrGlgcC5PmQMYIjbkYMsITIgfCVeHvCkpQFe1tQ$>:

> @@ -1,4 +1,26 @@
-function autoMerge(obj)
+function success = autoMerge(obj)


This looks (almost) identical to the autoMerge in DensityPeakClustering. We can move them both to Clustering instead, so changes later don't have to be duplicated.

________________________________

In +jrclust/+sort/@TemplateClustering/autoMerge.m<https://urldefense.com/v3/__https://github.com/JaneliaSciComp/JRCLUST/pull/144*discussion_r398569569__;Iw!!Eh6p8Q!WpOR-hIelz5ipAykGehbdqfmrGnn9mgVrGlgcC5PmQMYIjbkYMsITIgfCVeHvCm1QqmN2g$>:

> +fprintf( "autoMerge for TemplateClustering\n");
+if obj.nEdits > obj.editPos % not at tip of edit history, back out
+    warning('cannot branch from history; use revert() first');
+    return;
+end
+
+obj.orderClusters('clusterSites');
+
+for iRepeat = 1:obj.hCfg.nPassesMerge % single-pass vs dual-pass correction
+    nMerged = obj.mergeBySim();
+
+    if nMerged < 1
+        break;
+    end
+end
+fprintf( 'size of waveform similarity matrix: \n')


Again, something I'd log rather than print. Or is this just debug output?


=> Just debug. Removed.

________________________________

In +jrclust/+sort/@TemplateClustering/mergeBySim.m<https://urldefense.com/v3/__https://github.com/JaneliaSciComp/JRCLUST/pull/144*discussion_r398578186__;Iw!!Eh6p8Q!WpOR-hIelz5ipAykGehbdqfmrGnn9mgVrGlgcC5PmQMYIjbkYMsITIgfCVeHvClImMQVVg$>:

> @@ -1,5 +1,42 @@
 function nMerged = mergeBySim(obj)
-    %MERGEBYSIM
-    nMerged = 0;
+


This one is also (almost) identical to DensityPeakClustering.mergeBySim. I suggest moving them both to Clustering.


=> Let's keep here. I aspire to make it a choice to merge by waveform sim or template sim (which matches curation practice in phy), and that wouldn't apply for DensityPeakClustering.

________________________________

In +jrclust/+sort/@TemplateClustering/mergeBySim.m<https://urldefense.com/v3/__https://github.com/JaneliaSciComp/JRCLUST/pull/144*discussion_r398578435__;Iw!!Eh6p8Q!WpOR-hIelz5ipAykGehbdqfmrGnn9mgVrGlgcC5PmQMYIjbkYMsITIgfCVeHvCnCUsFVmw$>:

> @@ -1,5 +1,42 @@
 function nMerged = mergeBySim(obj)
-    %MERGEBYSIM
-    nMerged = 0;
+
+    fprintf('Template clustering mergeBySim\n');


Debug output?


=> Just debug. Removed.

________________________________

In +jrclust/+utils/metaToStruct.m<https://urldefense.com/v3/__https://github.com/JaneliaSciComp/JRCLUST/pull/144*discussion_r398579593__;Iw!!Eh6p8Q!WpOR-hIelz5ipAykGehbdqfmrGnn9mgVrGlgcC5PmQMYIjbkYMsITIgfCVeHvClni-9C6g$>:

> @@ -11,7 +11,7 @@
     end

     fid = fopen(filename, 'r');
-    keysvals = textscan(fid, '%s%s', 'Delimiter', '=',  'ReturnOnError', 0);
+    keysvals = textscan(fid, '%[^=] = %[^\r\n]',  'ReturnOnError', 0);


Is this more robust? Did the other way fail on you in some way? Does it handle cases where the = isn't separated by spaces?


=> More robust. In particular, when catGT is run, the command line (which has many equals signs in it) is written to the metadata and causes earlier parsing to fail. Works fine with our without spaces.

—


You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/JaneliaSciComp/JRCLUST/pull/144*pullrequestreview-381979100__;Iw!!Eh6p8Q!WpOR-hIelz5ipAykGehbdqfmrGnn9mgVrGlgcC5PmQMYIjbkYMsITIgfCVeHvCmTeuod2w$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AGVNOX2WCW3JVRP7LMVNBY3RJNLRHANCNFSM4LTBZNBQ__;!!Eh6p8Q!WpOR-hIelz5ipAykGehbdqfmrGnn9mgVrGlgcC5PmQMYIjbkYMsITIgfCVeHvCkfkFfwPA$>.

aliddell commented:
Okay! Ready to merge?

aliddell commented:
Wait, one small point: add yourself to the "contributors" array in `json/jrc.json`! I'm sorry I never added you before.

jenniferColonell commented:
thanks, Alan. Added me in alphabetical order.

ready to merge.

________________________________
From: Alan Liddell <notifications@github.com>
Sent: Thursday, March 26, 2020 12:08 PM
To: JaneliaSciComp/JRCLUST <JRCLUST@noreply.github.com>
Cc: Colonell, Jennifer <colonellj@janelia.hhmi.org>; Author <author@noreply.github.com>
Subject: Re: [JaneliaSciComp/JRCLUST] small fixes + add automerge for TemplateClustering (#144)


@aliddell approved this pull request.

Don't forget to add yourself to the contributors array!

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/JaneliaSciComp/JRCLUST/pull/144*pullrequestreview-382165304__;Iw!!Eh6p8Q!Xd5vHogtMnPJCDnYY_F86yKfZb24-GGtUIhaUYxpagfXujFmH7AF_sWGIv0lTuhNSGAXGQ$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AGVNOXYEZLA6Q5SBADMNLEDRJN4XTANCNFSM4LTBZNBQ__;!!Eh6p8Q!Xd5vHogtMnPJCDnYY_F86yKfZb24-GGtUIhaUYxpagfXujFmH7AF_sWGIv0lTuiQJ6pLIg$>.
